oak/server/rust/oak_runtime/src/node/http/mod.rs

```diff
@@ -0,0 +1,520 @@
+                    }
+                }))
+            }
+        });
+
+        let server = Server::bind(&self.address).serve(make_service);
+        let graceful = server.with_graceful_shutdown(async {
+            // Treat notification failure the same as a notification.
+            let _ = notify_receiver.await;
+        });
+        info!(
+            "{:?}: Started HTTP server pseudo-node on port {:?}",
+            std::thread::current().id(),
+            &self.address.port()
+        );
+
+        // Run until asked to terminate...
+        let result = graceful.await;
+        info!("HTTP server pseudo-node terminated with {:?}", result);
+    }
+
+    async fn map_to_http_request(req: Request<Body>) -> HttpRequest {
+        let url = req.uri().to_string();
+        let method = req.method().as_str().to_string();
+        let req_headers = req.headers();
+        let headers = req_headers
+            .into_iter()
+            .map(|(key, value)| {
+                let val = value.as_bytes().to_vec();
+                (key.to_string(), val)
+            })
+            .collect();
+        let body_stream = req.into_body();
+
+        let body = hyper::body::to_bytes(body_stream)
+            .await
+            .expect("Error when reading request body from the connection")
+            .to_vec();
+
+        HttpRequest {
+            url,
+            method,
+            body,
+            headers,
+        }
+    }
+}
+
+/// Oak Node implementation for the HTTP server.
+impl Node for HttpServerNode {
+    fn run(
+        self: Box<Self>,
+        runtime: RuntimeProxy,
+        startup_handle: oak_abi::Handle,
+        notify_receiver: oneshot::Receiver<()>,
+    ) {
+        // At start-of-day we need/expect to receive a write handle for an invocation channel
+        // to use for all subsequent activity.
+        info!("{}: Waiting for invocation channel", self.node_name);
+        let invocation_channel =
+            match HttpServerNode::try_get_invocation_channel(&runtime, startup_handle) {
+                Ok(writer) => writer,
+                Err(status) => {
+                    error!(
+                        "Failed to retrieve invocation channel write handle: {:?}",
+                        status
+                    );
+                    return;
+                }
+            };
+        if let Err(err) = runtime.channel_close(startup_handle) {
+            error!(
+                "Failed to close initial inbound channel {}: {:?}",
+                startup_handle, err
+            );
+        }
+
+        // Build a service to process all incoming HTTP requests.
+        let generic_handler = HttpRequestHandler {
+            runtime,
+            invocation_channel,
+        };
+        let server = self.make_server(generic_handler, notify_receiver);
+
+        // Create an Async runtime for executing futures.
+        // https://docs.rs/tokio/
+        let mut async_runtime = create_async_runtime();
```

Comment:
do we have any plans / github issue to clean up all these extra async runtimes? if not, could you create one and add a TODO here (and elsewhere)?

Reply:
The gRPC server node also creates a tokio runtime (and the auxiliary servers as well). It might be useful to have a single shared tokio runtime (perhaps as a field on the Oak runtime) that can be used by any server pseudo nodes that need to spawn things. Multiple independent runtimes probably lead to worse resource usage. This might also help with the nested runtime issue in the tests.